rust/worker/src/execution/orchestration/sparse_knn.rs

```diff
@@ -0,0 +1,267 @@
+use async_trait::async_trait;
+use chroma_blockstore::provider::BlockfileProvider;
+use chroma_error::{ChromaError, ErrorCodes};
+use chroma_system::{
+    wrap, ChannelError, ComponentContext, ComponentHandle, Dispatcher, Handler, Orchestrator,
+    OrchestratorContext, PanicError, TaskError, TaskMessage, TaskResult,
+};
+use chroma_types::{
+    operator::{Merge, RecordMeasure},
+    CollectionAndSegments, SparseVector,
+};
+use thiserror::Error;
+use tokio::sync::oneshot::{error::RecvError, Sender};
+use tracing::Span;
+
+use crate::execution::{
+    operators::{
+        sparse_index_knn::{
+            SparseIndexKnn, SparseIndexKnnError, SparseIndexKnnInput, SparseIndexKnnOutput,
+        },
+        sparse_knn_merge::{SparseKnnMergeError, SparseKnnMergeInput, SparseKnnMergeOutput},
+        sparse_log_knn::{SparseLogKnn, SparseLogKnnError, SparseLogKnnInput, SparseLogKnnOutput},
+    },
+    orchestration::knn_filter::KnnFilterOutput,
+};
+
+#[derive(Error, Debug)]
+pub enum SparseKnnError {
+    #[error("Operation aborted because resources exhausted")]
+    Aborted,
+    #[error("Error sending message through channel: {0}")]
+    Channel(#[from] ChannelError),
+    #[error("Panic: {0}")]
+    Panic(#[from] PanicError),
+    #[error("Error receiving final result: {0}")]
+    Result(#[from] RecvError),
+    #[error("Error running SparseKnnMerge operator: {0}")]
+    SparseKnnMerge(#[from] SparseKnnMergeError),
+    #[error("Error running SparseIndexKnn operator: {0}")]
+    SparseIndexKnn(#[from] SparseIndexKnnError),
+    #[error("Error running SparseLogKnn operator: {0}")]
+    SparseLogKnn(#[from] SparseLogKnnError),
+}
+
+impl ChromaError for SparseKnnError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            SparseKnnError::Aborted => ErrorCodes::ResourceExhausted,
+            SparseKnnError::Channel(err) => err.code(),
+            SparseKnnError::Panic(_) => ErrorCodes::Aborted,
+            SparseKnnError::Result(_) => ErrorCodes::Internal,
+            SparseKnnError::SparseKnnMerge(err) => err.code(),
+            SparseKnnError::SparseIndexKnn(err) => err.code(),
+            SparseKnnError::SparseLogKnn(err) => err.code(),
+        }
+    }
+}
+
+impl<E> From<TaskError<E>> for SparseKnnError
+where
+    E: Into<SparseKnnError>,
+{
+    fn from(value: TaskError<E>) -> Self {
+        match value {
+            TaskError::Aborted => SparseKnnError::Aborted,
+            TaskError::Panic(e) => e.into(),
+            TaskError::TaskFailed(e) => e.into(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SparseKnnOrchestrator {
```

Comment:
Lets sit down with Sanket next week, and map this to a cleaner I/O plan based on how we do SPANN